### PR TITLE
Random challenges - intro

### PR DESCRIPTION
**Pull request by patricia-ternes**
Introduction + 2 problems

I will add at least 3 more random number problems.

### CODE REVIEW

File: book/docs/random/coins-develop.ipynb [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    "def coin_flipper():\n",
+    "    coin = choice(['head','tail'])\n",
+    "    return coin"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Probability calculator\n",
```
Comment (Sparrow0hawk) [inline book/docs/random/coins-develop.ipynb:50]: I would explicitly state here that they should use the `coin_flipper` function for generating the probabilities. I initially read this and wrote:

``` 
def get_prob(toss1, toss2 = ''):

  choices = ['head','tail']

  prob = sum([toss1 == x for x in choices]) / len(choices)

  if toss2:
    toss2_prob = sum([toss2 == x for x in choices]) / len(choices)

    prob = prob * toss2_prob

  return prob
```

But realise on re-reading this isn't what you're after.

File: book/docs/random/coins-develop.ipynb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import unittest\n",
+    "\n",
+    "class UnitTests(unittest.TestCase):\n",
+    "    def test_coin_type(self):\n",
+    "        self.assertEqual(type(coin_flipper()), type('10'), 'The function should return a string')\n",
```
Comment (Sparrow0hawk) [inline book/docs/random/coins-develop.ipynb]:
```suggestion
    "        self.assertTrue(isinstance(type(coin_flipper()), str), 'The function should return a string')\n",
```
My understanding is the preferred way to check types is to use [`isinstance`](https://docs.python.org/3/library/functions.html#isinstance).

File: book/docs/random/coins-develop.ipynb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    "    def test_coin_head_tail(self):\n",
+    "        self.assertTrue(sorted(set(coin_flipper() for _ in range(100000))) == ['head', 'tail'], 'The function should return `head` or `tail`.')\n",
+    "    def test_coin_fairness(self):\n",
+    "        counts = [coin_flipper() for _ in range(100000)].count('head')/100000\n",
```
Comment (Sparrow0hawk) [inline book/docs/random/coins-develop.ipynb:98]: Did not know that `list` objects had a `.count` function! This is very cool! 🎉

File: book/docs/random/coins-solution.ipynb [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import unittest\n",
+    "\n",
+    "class UnitTests(unittest.TestCase):\n",
+    "    def test_coin_type(self):\n",
+    "        self.assertEqual(type(coin_flipper()), type('10'), 'The function should return a string')\n",
```
Comment (Sparrow0hawk) [inline book/docs/random/coins-solution.ipynb]: Same as above re using `isinstance`.

File: book/docs/random/dice-develop.ipynb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import unittest\n",
+    "from scipy import stats\n",
+    "import numpy as np\n",
+    "\n",
+    "class UnitTests(unittest.TestCase):\n",
+    "    def test_type(self):\n",
+    "        self.assertEqual(type(dice_rolling()), type(10), 'The function should return an integer')\n",
```
Comment (Sparrow0hawk) [inline book/docs/random/dice-develop.ipynb]: Same as above re using `isinstance` to test for type equality.

**Review by Sparrow0hawk — APPROVED** [review]
Looks good to me overall, i've made a couple of suggested changes for your tests but they are optional as the tests work!

File: book/docs/random/coins-develop.ipynb (hunk 2 above)
Comment (patricia-ternes) [reply book/docs/random/coins-develop.ipynb]: This way is much better! I will change!

File: book/docs/random/coins-develop.ipynb [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import unittest\n",
+    "\n",
+    "class UnitTests(unittest.TestCase):\n",
+    "    def test_coin_type(self):\n",
+    "        self.assertTrue(isinstance(type(coin_flipper()), str), 'The function should return a string')\n",
```
Comment (patricia-ternes) [inline book/docs/random/coins-develop.ipynb:94]:
```suggestion
    "        self.assertTrue(isinstance(coin_flipper(), str), 'The function should return a string')\n",
```

File: book/docs/random/coins-develop.ipynb [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    "    def test_coin_head_tail(self):\n",
+    "        self.assertTrue(sorted(set(coin_flipper() for _ in range(100000))) == ['head', 'tail'], 'The function should return `head` or `tail`.')\n",
+    "    def test_coin_fairness(self):\n",
+    "        counts = [coin_flipper() for _ in range(100000)].count('head')/100000\n",
+    "        self.assertAlmostEqual(counts, 0.5, places=2, msg=f'The side should be rolled with a probability of 0.50')\n",
+    "\n",
+    "    def test_probability_type(self):\n",
+    "        self.assertEqual(type(get_probability('head','head')), type(10.0), 'The function should return a float')\n",
```
Comment (patricia-ternes) [inline book/docs/random/coins-develop.ipynb:102]:
```suggestion
    "        self.assertTrue(isinstance(get_probability('head','head'), float), 'The function should return a float')\n",
```

File: book/docs/random/coins-solution.ipynb [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+    "    def test_coin_head_tail(self):\n",
+    "        self.assertTrue(sorted(set(coin_flipper() for _ in range(100000))) == ['head', 'tail'], 'The function should return `head` or `tail`.')\n",
+    "    def test_coin_fairness(self):\n",
+    "        counts = [coin_flipper() for _ in range(100000)].count('head')/100000\n",
+    "        self.assertAlmostEqual(counts, 0.5, places=2, msg=f'The side should be rolled with a probability of 0.50')\n",
+    "\n",
+    "    def test_probability_type(self):\n",
+    "        self.assertEqual(type(get_probability('head','head')), type(10.0), 'The function should return a float')\n",
```
Comment (patricia-ternes) [inline book/docs/random/coins-solution.ipynb:200]:
```suggestion
    "        self.assertTrue(isinstance(get_probability('head','head'), float), 'The function should return a float')\n",
```

File: book/docs/random/dice-develop.ipynb (hunk 5 above)
Comment (patricia-ternes) [reply book/docs/random/dice-develop.ipynb]:
```suggestion
    "        self.assertTrue(isinstance(dice_rolling(), int), 'The function should return an integer')\n",
```

File: book/docs/random/dice-solution.ipynb [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import unittest\n",
+    "import numpy as np\n",
+    "\n",
+    "class UnitTests(unittest.TestCase):\n",
+    "    def test_type(self):\n",
+    "        self.assertEqual(type(dice_rolling()), type(10), 'The function should return an integer')\n",
```
Comment (patricia-ternes) [inline book/docs/random/dice-solution.ipynb:72]:
```suggestion
    "        self.assertTrue(isinstance(dice_rolling(), int), 'The function should return an integer')\n",
```

File: book/docs/random/coins-solution.ipynb (hunk 4 above)
Comment (patricia-ternes) [reply book/docs/random/coins-solution.ipynb]:
```suggestion
    "        self.assertTrue(isinstance(coin_flipper(), str), 'The function should return a string')\n",
```

File: book/docs/random/coins.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
+# Tossing Coins 🟡
+
+> Write a function that works as a coin-flipper simulator.
+> Then, write another function that returns the approximate probability of
```
Comment (patricia-ternes) [inline book/docs/random/coins.md:4]:
```suggestion
> Then, write another function that calls the coin flipper function several
> times to determine the approximate probability of
```

File: book/docs/random/coins-develop.ipynb [hunk 11]
```diff
@@ -0,0 +1,148 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Write your code: Tossing Coins\n",
+    "\n",
+    "> Write a function that works as a coin-flipper simulator.\n",
+    "> Then, write another function that returns the approximate probability of\n",
```
Comment (patricia-ternes) [inline book/docs/random/coins-develop.ipynb:10]:
```suggestion
    "> Then, write another function that calls the coin flipper function several\n",
    "> times to determine the approximate probability of\n",
```

File: book/docs/random/coins-solution.ipynb [hunk 12]
```diff
@@ -0,0 +1,246 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Solution: Tossing Coins\n",
+    "\n",
+    "> Write a function that works as a coin-flipper simulator.\n",
+    "> Then, write another function that returns the approximate probability of\n",
```
Comment (patricia-ternes) [inline book/docs/random/coins-solution.ipynb:10]:
```suggestion
    "> Then, write another function that calls the coin flipper function several\n",
    "> times to determine the approximate probability of\n",
```

File: book/_toc.yml [hunk 13]
```diff
@@ -30,3 +30,14 @@ chapters:
...
+- file: docs/random/intro
+  sections:
+  - file: docs/random/dice
+    sections:
+    - file: docs/random/dice-develop
+    - file: docs/random/dice-solution
+  - file: docs/random/coins
+    sections:
+    - file: docs/random/coins-develop
+    - file: docs/random/coins-solution
```
Comment (patricia-ternes) [inline book/_toc.yml]:
```suggestion
    - file: docs/random/coins-solution

```

**Review by patricia-ternes — COMMENTED** [review]
Addressing @Sparrow0hawk 's comments.

**Pull request merged**